sql: set mutation columns to nullable while backfilling

vivekmenezes opened the pull request:
Also added the missing UPSERT tests.

The better way to fix this in the future is to always set
mutable columns to non-nullable in the yet to be introduced
ImmutableTableDescriptor (the opposite of MutableTableDescriptor)

fixes #32530

Release note (sql change): Fix crash in UPSERT in the middle of
a schema change adding a non-nullable column.

vivekmenezes committed:
sql: test UPSERT in TestRaceWithBackfill

Release note: None

vivekmenezes committed:
sql: set mutation columns to nullable while backfilling

Also added the missing UPSERT tests. Add tests for INSERT
and DELETE.

The better way to fix this in the future is to always set
mutable columns to non-nullable in the yet to be introduced
ImmutableTableDescriptor (the opposite of MutableTableDescriptor)

fixes #32530

Release note (sql change): Fix crash in UPSERT in the middle of
a schema change adding a non-nullable column.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32585)
<!-- Reviewable:end -->

vivekmenezes commented:
filled away future cleanup https://github.com/cockroachdb/cockroach/issues/32586

knz commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -1886,6 +1886,29 @@ func (desc *TableDescriptor) FindActiveColumnByID(id ColumnID) (*ColumnDescripto
+				// Even if the column is non-nullable it can be null in the
+				// middle of a schema change while it's being baxckfilled
+				// to a non-nullable value.
+				col.Nullable = true
```
@vivekmenezes I think we need to extend this change. Let me explain.

The change you are making here appears to help the row **reader** to not be confused when it sees a null value in a non-nullable column. I have some comments about this (see point 2 below) but let's keep it at that for now.

1) The main issue here is that changing this property will create a semantic problem for **row writers** -- it is important that a client app **receives a SQL error if they attempt to insert a NULL value into a non-NULLable column.** With your change here, I believe this check is not done any more.

So I think the first order of business is to verify (with tests) that NULL inserts into non-NULLable columns currently undergoing a backfill are still/properly rejected by SQL errors. 

2) regarding the point at the top, with the row readers. In any case it is invalid for a SQL client to receive a NULL value as response for a query to a non-NULL column. We must not allow this to happen in any case -- this is a plain correctness error, a silent data error (S-0). 

I wonder if your change here makes it more or less likely to encounter this situation. If the column is marked as nullable in the descriptor, and there is a query to read a row in the process of being added, the correct behavior is this:

   2.a) the query must not see the column at all -- it must operate on the table as if the column did not exist
   2.b) in any case, there must not be any NULL value visible by the query

In contrast, the following behavior would be incorrect:

   2.c) the query sees the column
   2.d) the query sees a NULL value in the column


What would you say in response to my comments here?

knz replied:
Also to ensure the new added column is not visible it will be interesting to ensure:

1) we have a test that the following statement groups each produce an error:

```sql
CREATE TABLE t (x INT PRIMARY KEY);
BEGIN;
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 123;
INSERT INTO t (x) VALUES (1) ON CONFLICT DO UPDATE SET x = t.y + 1; -- this must error with "t.y is not public"
COMMIT;
```

2) that the behavior with stored columns that depend on both new columns and existing columns is clear:

```sql
CREATE TABLE t (x INT PRIMARY KEY);
BEGIN;
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 123,
              ADD COLUMN z INT AS (x + y) STORED;
INSERT INTO t (x) VALUES (2); --  see note below
COMMIT;
```

In this example the column `z` must be updated both by the backfill _and_ the insert, so that the final value of `z` is properly equal to 125.

vivekmenezes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32585)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sqlbase/structured.go, line 1904 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSDH3kT87GH53jIYJDG-r2-1904:-LSGdlzY2Irs6bQi7SDc:b-3d9ziv) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b10a626a6834406821ca71dd55499580788978/pkg/sql/sqlbase/structured.go#L1904)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Also to ensure the new added column is not visible it will be interesting to ensure:

1) we have a test that the following statement groups each produce an error:

```sql
CREATE TABLE t (x INT PRIMARY KEY);
BEGIN;
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 123;
INSERT INTO t (x) VALUES (1) ON CONFLICT DO UPDATE SET x = t.y + 1; -- this must error with "t.y is not public"
COMMIT;
```

2) that the behavior with stored columns that depend on both new columns and existing columns is clear:

```sql
CREATE TABLE t (x INT PRIMARY KEY);
BEGIN;
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 123,
              ADD COLUMN z INT AS (x + y) STORED;
INSERT INTO t (x) VALUES (2); --  see note below
COMMIT;
```

In this example the column `z` must be updated both by the backfill _and_ the insert, so that the final value of `z` is properly equal to 125.
</blockquote></details>

I've beefed up the tests. PTAL. Thanks!


<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@knz I'd love to merge this soon so that we can declare this fixed! Thanks!

knz reviewed (commented):
Thank you for pinging me. I have reviewed this again, see my comments below.

> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-:-LSLI_qP46bC13QWlK3g:bk0fio9)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/schema_changer_test.go, line 2241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSLIAV0ERfCqfp1jSjn:-LSLIAV1DTLN_-10Njjh:bdk14uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e68cd83a2bfac1d94afc3d0c6bee8cfb5377d82/pkg/sql/schema_changer_test.go#L2241)):*
> ```Go
> 	var wg sync.WaitGroup
> 	wg.Add(1)
> 	go func() {
> ```

I like the SQL in this test but I do not understand the architecture of the test.

It seems to me like we want the test to issue the schema change concurrently with _each_ of the SQL statements below. Right now the test is non-deterministic and probably completes the schema change very early, before the first few SQL statements below are executed.

What confidence can we create that all the SQL statements below are executed concurrently with a schema change?

Is there a way to block the SC transaction and see all these statements:

A) assert that there is a mutation present in the descriptor (if there is no mutation, the test is not testing anything)

B) that despite the mutation the behavior is correct
___
*[pkg/sql/sqlbase/structured.go, line 1902 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSLFVTl56O5Raq4m1pd:-LSLFVTm3yRniv6-QOim:b-ej7zt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e68cd83a2bfac1d94afc3d0c6bee8cfb5377d82/pkg/sql/sqlbase/structured.go#L1902)):*
> ```Go
> 				col := *c
> 				// Even if the column is non-nullable it can be null in the
> 				// middle of a schema change while it's being baxckfilled
> ```

1) nit: typo

2) why do this in `FindReadableColumnById`? Is this because you only want the Nullable property to be set on a copy (so the desc is unaffected)?

If so, how can we be sure that the columns are not accessed in other ways than via `FindReadableColumnByID`?

In particular:

- `makeUpdaterWithoutCascader()` (`row/writer.go`) iterates on Mutations directly and is probably sensitive to having the Nullable bit set properly
- ditto deleter, etc
- `ProcessDefaultColumns` / `processColumnSet` (`default_exprs.go`) iterate on `Mutations` directly but need to have appropriate nullability information for computing default values


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
Thanks for taking a look. PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-:-LSLudWA6_l3UCpoAcia:b2riqia)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/schema_changer_test.go, line 2241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSLIAV0ERfCqfp1jSjn:-LSLJtbrDe0Y5rX_l2pQ:b-h3hbjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e68cd83a2bfac1d94afc3d0c6bee8cfb5377d82/pkg/sql/schema_changer_test.go#L2241)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I like the SQL in this test but I do not understand the architecture of the test.

It seems to me like we want the test to issue the schema change concurrently with _each_ of the SQL statements below. Right now the test is non-deterministic and probably completes the schema change very early, before the first few SQL statements below are executed.

What confidence can we create that all the SQL statements below are executed concurrently with a schema change?

Is there a way to block the SC transaction and see all these statements:

A) assert that there is a mutation present in the descriptor (if there is no mutation, the test is not testing anything)

B) that despite the mutation the behavior is correct
</blockquote></details>

The schema changer is blocked  from executing through testing knobs. I can do both A and B here. Keep in mind we do test behavior in many tests in descriptor_mutations_test.go. This test is for testing that none of these statements will crash. 

Added both A and B
___
*[pkg/sql/sqlbase/structured.go, line 1902 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSLFVTl56O5Raq4m1pd:-LSLo-r794o7F_wNiVJJ:b-upth4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e68cd83a2bfac1d94afc3d0c6bee8cfb5377d82/pkg/sql/sqlbase/structured.go#L1902)):*
> makeUpdaterWithoutCascader
I've modified this to introduce a FindInactiveColumnById() method and moved the code to set the  appropriate column to nullable to the writer.go code.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Thank you for the changes. The test looks good. 
I think some other test needs to be adapted (see CI failure)

LGTM modulo double-checking the rest of the code in `row/writer.go` - perhaps the deleter needs the same change.

> Reviewed 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-:-LSOmMO79dBzlAGsIgxH:b-qgnr27)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sqlbase/structured.go, line 1902 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSLFVTl56O5Raq4m1pd:-LSOlssA83XrdLlN3oA4:bf4da77) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e68cd83a2bfac1d94afc3d0c6bee8cfb5377d82/pkg/sql/sqlbase/structured.go#L1902)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

> makeUpdaterWithoutCascader
I've modified this to introduce a FindInactiveColumnById() method and moved the code to set the  appropriate column to nullable to the writer.go code.
</blockquote></details>

Thank you. Can you double check the other writers in the same file. I think also the deleter is affected.


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
The test failure is https://github.com/cockroachdb/cockroach/issues/32671

TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-:-LSPtGrC-_joQgm0_Jf0:b-6bcs5h)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sqlbase/structured.go, line 1902 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-LSLFVTl56O5Raq4m1pd:-LSPtKp2BEotshBdDgzB:bcsd2ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e68cd83a2bfac1d94afc3d0c6bee8cfb5377d82/pkg/sql/sqlbase/structured.go#L1902)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Thank you. Can you double check the other writers in the same file. I think also the deleter is affected.
</blockquote></details>

We do have DELETE in the unittests. The reason why DELETE is not affected is because it doesn't read any rows when performing its operations, and when using an expression it can only use columns that are public.


<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@knz I don't believe we allow such FK relationships to exist. Let me add a test for that

vivekmenezes committed:
sql: test a FK constraint cannot be added over an inactive column

Release note: None

vivekmenezes committed:
sql: allow ADD NOT NULL computed columns

Release note (sql change): fix error when configuring NOT NULL
computed columns.

vivekmenezes commented:
@knz I've added two more commits to this PR. The first one fixes a minor bug and the other checks that we cannot add FK relationships over columns that are inactive. PTAL

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work! :sparkling_heart: 

> Reviewed 3 of 3 files at r8, 2 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32585#-:-LSVNWUyAxWOWcSoGW3W:b-ify0fq)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

vivekmenezes commented:
TFTR!

vivekmenezes commented:
bors r+
